### DISCUSSION TITLE
Component CSS and JavaScript always loading

### DISCUSSION THREAD

**Discussion by bstoots** [django-components/django-components#70] [Q&A] [answered]
Hey gang, I'm running into some unexpected behavior whereby component CSS and JavaScript is always being loaded even if the component is not being used on the page.  In fact it's injecting these into every page of my app.

When I remove the `@component.register` decorator from my component class it is no longer injected anywhere so it doesn't seem like a weird caching issue or anything.

Any thoughts on how I might go about tracking down this issue?  My version is `0.14`.

**Comment by EmilStenstrom** [comment]
Oh, that's missing documentation, sorry about that. You need to add 'django_components.middleware.ComponentDependencyMiddleware' to your MIDDLEWARE section too, that's the thing that makes smarter choices of where to import CSS and JS.

I'll create an issue about adding docs for this.

**Reply by bstoots** [reply]
Much obliged!  I'll stand by for the official docs but from an initial attempt it doesn't look like adding `django_components.middleware.ComponentDependencyMiddleware` solved my problem.  I am also using WhiteNoise for serving static assets so perhaps there's a weird interaction there.  I'll see if I can hone in on it.

**Reply by rbeard0330** [reply]
The middleware-based renderer is actually hidden behind another setting while it's in undocumented state.  Try adding:
```
'COMPONENTS' {
    'RENDER_DEPENDENCIES': True
}
```
to your settings and see if that improves the situation.

**Reply by bstoots** [reply]
> 
> 
> The middleware-based renderer is actually hidden behind another setting while it's in undocumented state. Try adding:
> 
> ```
> 'COMPONENTS' {
>     'RENDER_DEPENDENCIES': True
> }
> ```
> 
> to your settings and see if that improves the situation.

Adding `RENDER_DEPENDENCIES` didn't seem to change anything.  All component CSS and JS is still loaded.  Toggling that key between True and False doesn't seem to do anything either, what is the intent of that key?

**Reply by rbeard0330** [reply]
> > The middleware-based renderer is actually hidden behind another setting while it's in undocumented state. Try adding:
> > ```
> > 'COMPONENTS' {
> >     'RENDER_DEPENDENCIES': True
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > to your settings and see if that improves the situation.
> 
> Adding `RENDER_DEPENDENCIES` didn't seem to change anything. All component CSS and JS is still loaded. Toggling that key between True and False doesn't seem to do anything either, what is the intent of that key?

There are two ways to include JS/CSS.  The original approach was to render all tags for all components where the relevant template tag appears.  The new approach is to have those tags just leave a placeholder HTML comment, and to also have all of the components leave similar comments for the dependencies they need.  Then the middleware looks at the completed response strips out all the comments, and puts the requested dependencies where the placeholders were.

All that behavior is governed by calling the `RENDER_DEPENDENCIES` setting--if it's false, the old behavior is used, and if it's true, it's the new behavior.

**Reply by EmilStenstrom** [reply]
@rbeard0330 I guess we should make an issue to make the new way the default way? Since it's clearly better, and should be backwards compatible?

**Reply by EmilStenstrom** [reply]
@bstoots If there's a bug that prevents things from working correctly more trouble shooting information would be helpful. 

For reference (since there's not much docs yet) here's how we expect things to work in our tests: https://github.com/EmilStenstrom/django-components/blob/master/tests/test_dependency_rendering.py#L47

**Reply by bstoots** [reply]
> 
> 
> @bstoots If there's a bug that prevents things from working correctly more trouble shooting information would be helpful.
> 
> For reference (since there's not much docs yet) here's how we expect things to work in our tests: https://github.com/EmilStenstrom/django-components/blob/master/tests/test_dependency_rendering.py#L47

I fiddled around with it a bit this morning and didn't find any resolution to my problem.  I'll have to stand up a clean application and see if I can isolate the cause but it will probably be at least a few days until I have time to look into it.

**Reply by EmilStenstrom** [reply]
@bstoots Sounds great! We'd love to help, it's just hard to guess what's wrong in each case :)

**Reply by rafae56038** [reply, marked as answer]
Using this in the settings of project solves it.



`
COMPONENTS = {
    'RENDER_DEPENDENCIES': True
}
`
